samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/BeginDeliveryActionListener.java

```diff
@@ -453,10 +455,22 @@ else if (minute > 1) {
     		if (extTimeService.hasExtendedTime() && extTimeService.getTimeLimit() > 0) {
     			control.setTimeLimit(extTimeService.getTimeLimit());
     		}
-    		String timeLimitInSetting = control.getTimeLimit() == null ? "0" : control.getTimeLimit().toString();
     		Date attemptDate = unSubmittedAssessmentGrading.getAttemptDate();
+    		long timeLimitInSetting = control.getTimeLimit();
+    		Long now = new Date().getTime();
+    		Long start = attemptDate.getTime();
+    		if((now - start) > (timeLimitInSetting*1000)) {
+    			// check if the queue is ahead and already submitted it
+    			TimedAssessmentQueue queue = TimedAssessmentQueue.getInstance();
+    			TimedAssessmentGradingModel timedAG = (TimedAssessmentGradingModel)queue.
+    					get(unSubmittedAssessmentGrading.getAssessmentGradingId());
+    			// if it was submitted (race condition) while checking, unblock it - sam will synch soon
+    			if(timedAG != null && !timedAG.getSubmittedForGrade()) {
+    				delivery.setTimeExpired(true);
+    			}
+    		}
     		delivery.setBeginTime(attemptDate);
-    		String timeBeforeDueRetract = delivery.getTimeBeforeDueRetract(timeLimitInSetting);
+    		String timeBeforeDueRetract = delivery.getTimeBeforeDueRetract(control.getTimeLimit() == null ? "0" : String.valueOf(control.getTimeLimit()));
```

Comment:
Hey @bjones86 , what I meant is that because the arg for getTimeBeforeDueRetract has to be a String, I will be casting into a String the Integer I get from getTimeLimit, pass it to NumberUtils and then, after number utils returns an integer, I need to cast it yet again to a String. I have no problem but it seems a lot of cast back and forth when in the original call I am casting only once.

Thanks.